### PR TITLE
fix(drive-abci): bound CheckTx verification work

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Summary

- perform current-state admission before shielded cryptographic validation in CheckTx
- add a small action-weighted CheckTx verification budget isolated from proposal and block processing
- return transient backpressure when local verification capacity is occupied

Addresses security review findings DS-CAND-207 through DS-CAND-211.

## Verification

- `cargo test -p drive-abci --lib check_tx_verification -- --nocapture`
- focused proof-budget and transport-status regression tests
- `cargo clippy -p drive-abci --lib --no-deps -- -D warnings`
- `cargo fmt --all -- --check`
- `git diff --check`

### CODE REVIEW

**Comment by thepastaclaw** [comment]
✅ Final review complete — no blockers (commit f7a88430)

File: packages/rs-drive-abci/src/execution/validation/state_transition/check_tx_verification/v0/mod.rs [hunk 1]
```diff
@@ -336,6 +325,25 @@ pub(super) fn state_transition_to_execution_event_for_check_tx_v0<'a, C: CoreRPC
...
+            // Run proof verification only after cheap current-state admission has
+            // accepted the transition. CheckTx uses a node-local, fail-fast budget;
+            // proposal and block execution do not share it and retain capacity.
+            if state_transition.has_shielded_proof_validation() {
+                let _permit = proof_verifier
+                    .try_acquire(state_transition.shielded_proof_action_count())
+                    .ok_or(Error::Execution(
+                        ExecutionError::CheckTxProofVerificationBusy,
+                    ))?;
+                let result = state_transition.validate_shielded_proof(platform_version)?;
```
Comment (thepastaclaw) [inline packages/rs-drive-abci/src/execution/validation/state_transition/check_tx_verification/v0/mod.rs:373]: **🔴 Blocking: ShieldFromAssetLock bypasses the CheckTx proof-verification budget**

The permit is conditional on has_shielded_proof_validation(), which explicitly excludes ShieldFromAssetLock. That transition's earlier transform_into_action call instead invokes reconstruct_and_verify_bundle directly at shield_from_asset_lock/transform_into_action/v0/mod.rs:298. A bad proof returns a PartiallyUseAssetLockAction together with a consensus error, but this CheckTx path converts the result to errors only at lines 318-323, and check_tx_v0 rejects it without executing or persisting the penalty action. The asset-lock owner can therefore mutate a canonical proof, re-sign each transition, and repeatedly run Halo2 verification against the same funded lock outside the new action-weighted limit. Acquire the CheckTx permit around ShieldFromAssetLock proof verification after its cheap asset-lock and state admission checks, while preserving the transformer's penalty action for proposal and block processing, and add a regression test covering this path.

<sub>source: ['codex']</sub>

File: packages/rs-drive-abci/src/platform_types/check_tx_proof_verifier.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        let verifier = CheckTxProofVerifier::new(3);
+        let first = verifier.try_acquire(1).expect("first permit");
+        let second = verifier.try_acquire(4).expect("weighted second permit");
+
+        assert!(verifier.try_acquire(1).is_none());
+
+        drop(first);
+        assert!(verifier.try_acquire(1).is_some());
+
+        drop(second);
```
Comment (thepastaclaw) [inline packages/rs-drive-abci/src/platform_types/check_tx_proof_verifier.rs:101]: **🟡 Suggestion: Exercise the limiter under concurrent contention**

The current test acquires permits sequentially and never sends competing callers through the compare-exchange retry path. Add a barrier-based multithreaded test that holds admitted permits, records the maximum simultaneous weight, verifies it never exceeds the configured limit, and confirms all capacity is released afterward. This directly protects the concurrency property the limiter was introduced to enforce.

<sub>source: ['codex']</sub>

File: packages/rs-drive-abci/src/abci/handler/check_tx.rs [hunk 3]
```diff
@@ -148,6 +148,13 @@ where
             })
         })
         .or_else(|error| {
+            if matches!(
+                error,
+                Error::Execution(ExecutionError::CheckTxProofVerificationBusy)
+            ) {
+                return Err(error);
+            }
```
Comment (thepastaclaw) [inline packages/rs-drive-abci/src/abci/handler/check_tx.rs:156]: **🟡 Suggestion: Cover backpressure through the complete CheckTx error path**

Retryable backpressure depends on the validator producing CheckTxProofVerificationBusy, this handler propagating it instead of converting it into an in-band ABCI response, and the application mapping it to tonic ResourceExhausted. Existing tests cover the validator and status mapper independently, so removing this special case would still leave them green. Add a handler/application-level test that occupies the permit, submits a state-admissible shielded transition, and asserts the resulting tonic status is ResourceExhausted.

<sub>source: ['codex']</sub>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Preliminary review — Codex only

Verified the full PR delta at exact head 0d5219e9a3ba3401404a807941703cbb6b05a973. The limiter correctly protects the shielded proofs routed through validate_shielded_proof, and the focused tests pass, but ShieldFromAssetLock still performs Halo2 verification before the permit and CheckTx discards its penalty action on failure. This leaves a repeatable CPU-exhaustion path, so changes are required; two non-blocking regression-test gaps are also noted.

Validated blockers were found in the Codex precheck. Sonnet is deferred until a fresh Codex revalidation clears the blocker gate.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (failed), `gpt-5.6-sol` — security-auditor (failed), `gpt-5.6-sol` — rust-quality (failed), `gpt-5.6-sol` — general (failed), `gpt-5.6-sol` — security-auditor (failed), `gpt-5.6-sol` — rust-quality (failed), `gpt-5.6-sol` — general (failed), `gpt-5.6-sol` — security-auditor (failed), `gpt-5.6-sol` — rust-quality (failed), `gpt-5.6-sol` — general (failed), `gpt-5.6-sol` — security-auditor (failed), `gpt-5.6-sol` — rust-quality (failed), `gpt-5.6-sol` — general (failed), `gpt-5.6-sol` — security-auditor (failed), `gpt-5.6-sol` — rust-quality (failed), `gpt-5.6-sol` — general (failed), `gpt-5.6-sol` — security-auditor (failed), `gpt-5.6-sol` — rust-quality (failed), `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — security-auditor (completed), `gpt-5.6-sol` — rust-quality (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet: **not run (deferred by blocker gate)**

🔴 1 blocking | 🟡 2 suggestion(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-drive-abci/src/execution/validation/state_transition/check_tx_verification/v0/mod.rs`:
- [BLOCKING] packages/rs-drive-abci/src/execution/validation/state_transition/check_tx_verification/v0/mod.rs:331-337: ShieldFromAssetLock bypasses the CheckTx proof-verification budget
  The permit is conditional on has_shielded_proof_validation(), which explicitly excludes ShieldFromAssetLock. That transition's earlier transform_into_action call instead invokes reconstruct_and_verify_bundle directly at shield_from_asset_lock/transform_into_action/v0/mod.rs:298. A bad proof returns a PartiallyUseAssetLockAction together with a consensus error, but this CheckTx path converts the result to errors only at lines 318-323, and check_tx_v0 rejects it without executing or persisting the penalty action. The asset-lock owner can therefore mutate a canonical proof, re-sign each transition, and repeatedly run Halo2 verification against the same funded lock outside the new action-weighted limit. Acquire the CheckTx permit around ShieldFromAssetLock proof verification after its cheap asset-lock and state admission checks, while preserving the transformer's penalty action for proposal and block processing, and add a regression test covering this path.

In `packages/rs-drive-abci/src/platform_types/check_tx_proof_verifier.rs`:
- [SUGGESTION] packages/rs-drive-abci/src/platform_types/check_tx_proof_verifier.rs:87-98: Exercise the limiter under concurrent contention
  The current test acquires permits sequentially and never sends competing callers through the compare-exchange retry path. Add a barrier-based multithreaded test that holds admitted permits, records the maximum simultaneous weight, verifies it never exceeds the configured limit, and confirms all capacity is released afterward. This directly protects the concurrency property the limiter was introduced to enforce.

In `packages/rs-drive-abci/src/abci/handler/check_tx.rs`:
- [SUGGESTION] packages/rs-drive-abci/src/abci/handler/check_tx.rs:150-156: Cover backpressure through the complete CheckTx error path
  Retryable backpressure depends on the validator producing CheckTxProofVerificationBusy, this handler propagating it instead of converting it into an in-band ABCI response, and the application mapping it to tonic ResourceExhausted. Existing tests cover the validator and status mapper independently, so removing this special case would still leave them green. Add a handler/application-level test that occupies the permit, submits a state-admissible shielded transition, and asserts the resulting tonic status is ResourceExhausted.
```
</details>

File: packages/rs-drive-abci/src/execution/validation/state_transition/check_tx_verification/v0/mod.rs (hunk 1 above)
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/validation/state_transition/check_tx_verification/v0/mod.rs:373]: Addressed in f7a884301f. ShieldFromAssetLock now acquires the same bounded proof-verification permit after cheap admission checks and before expensive verification. Focused CheckTx tests and strict clippy pass.

File: packages/rs-drive-abci/src/platform_types/check_tx_proof_verifier.rs (hunk 2 above)
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/platform_types/check_tx_proof_verifier.rs:101]: Addressed in f7a884301f. A barrier-based concurrent contention test verifies that the configured limiter admits only the bounded number of simultaneous proof verifications.

File: packages/rs-drive-abci/src/abci/handler/check_tx.rs (hunk 3 above)
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/abci/handler/check_tx.rs:156]: Addressed in f7a884301f. The handler/application integration test now drives limiter saturation through the complete CheckTx path and asserts RESOURCE_EXHAUSTED behavior.
